File: src/services/customer-image-label.service.ts
import { CustomerImageLabelRepository } from "@/repositories/customer-image-label.repository";
import type { CustomerImageLabel } from "@/types/customer/customer_image_label";
import { AppError } from "@/middleware/error-handler";
import { logger } from "@/utils/logger";

export class CustomerImageLabelService {
  private repository: CustomerImageLabelRepository;

  constructor() {
    this.repository = new CustomerImageLabelRepository();
  }

  async createLabel(companyId: string, name: string): Promise<CustomerImageLabel> {
    // Omit _id - MongoDB will auto-generate ObjectId
    const label: Omit<CustomerImageLabel, "_id"> = {
      name,
    };

    return await this.repository.create(companyId, label);
  }

  async getLabel(companyId: string, id: string): Promise<CustomerImageLabel> {
    const label = await this.repository.findById(companyId, id);

    if (!label) {
      throw new AppError(404, "Label not found", "LABEL_NOT_FOUND");
    }

    return label;
  }

  async listLabels(companyId: string): Promise<CustomerImageLabel[]> {
    return await this.repository.findAll(companyId);
  }

  async updateLabel(
    companyId: string,
    id: string,
    name: string
  ): Promise<CustomerImageLabel> {
    const exists = await this.repository.exists(companyId, id);
    if (!exists) {
      throw new AppError(404, "Label not found", "LABEL_NOT_FOUND");
    }

    const updated = await this.repository.update(companyId, id, { name });

    if (!updated) {
      throw new AppError(500, "Failed to update label", "UPDATE_FAILED");
    }

    return updated;
  }

  async deleteLabel(companyId: string, id: string): Promise<void> {
    const exists = await this.repository.exists(companyId, id);
    if (!exists) {
      throw new AppError(404, "Label not found", "LABEL_NOT_FOUND");
    }

    // Use transaction to delete label and remove from all images
    const deleted = await this.repository.deleteWithTransaction(companyId, id);

    if (!deleted) {
      throw new AppError(500, "Failed to delete label", "DELETE_FAILED");
    }

    logger.info("Label deleted successfully", { id, companyId });
  }
}

File: fisoft/src/main/kotlin/services/CustomerImageLabelService.kt
package com.ilyasdemirkiran.services

import com.ilyasdemirkiran.middleware.AppError
import com.ilyasdemirkiran.repository.CustomerImageLabelRepository
import com.ilyasdemirkiran.types.customers.CustomerImageLabel
import io.ktor.http.*

class CustomerImageLabelService(
    private val repo: CustomerImageLabelRepository = CustomerImageLabelRepository()
) {
    suspend fun createLabel(companyId: String, name: String): CustomerImageLabel {
        return repo.create(companyId, name)
    }

    suspend fun getLabel(companyId: String, id: String): CustomerImageLabel {
        return repo.findById(companyId, id)
            ?: throw AppError(HttpStatusCode.NotFound, "Label not found", "LABEL_NOT_FOUND")
    }

    suspend fun listLabels(companyId: String): List<CustomerImageLabel> {
        return repo.findAll(companyId)
    }

    suspend fun updateLabel(companyId: String, id: String, name: String): CustomerImageLabel {
        return repo.update(companyId, id, name)
            ?: throw AppError(HttpStatusCode.NotFound, "Label not found", "LABEL_NOT_FOUND")
    }

    suspend fun deleteLabel(companyId: String, id: String) {
        if (!repo.delete(companyId, id)) {
            throw AppError(HttpStatusCode.NotFound, "Label not found", "LABEL_NOT_FOUND")
        }
    }
}
